src/loading.ts

```diff
@@ -0,0 +1,73 @@
+import { OperatorFunction, merge, combineLatest, of } from 'rxjs'
+import { share, startWith, map, filter, mapTo, delay, endWith, scan } from 'rxjs/operators'
+import { isEqual } from 'lodash'
+
+export const LOADING = 'loading' as const
+
+/**
+ * An emission from a result provider.
+ *
+ * @template T The type of the result. Should typically include an empty value, or even an error type.
+ */
+export interface MaybeLoadingResult<T> {
+    /**
+     * Whether the result provider is currently getting a new result.
+     */
+    isLoading: boolean
+
+    /**
+     * The latest result.
+     */
+    result: T
+}
```

Comment:
If I understood correctly the implementation of `emitLoading()` below, emissions where `isLoading === false && result !== emptyResultValue` will get ignored. Should we represent this at the type level? Eg.:

```typescript
export type MaybeLoadingResult<TResult, TEmpty> = { isLoading: boolean; result: TEmpty } | { isLoading: false; result: TResult }
```

Reply:
I think that would disallow intermediate results from the provider - e.g. in the webapp, multiple providers are invoked in parallel, and whenever one of them returns there will be an `{ isLoading: true, result: TResult }` emission as long as there are still other providers loading. At first I thought that you are right though, and that reading the logic this is (wrongly) ignored. There was no unit test for intermediate values, so I added one in https://github.com/sourcegraph/codeintellify/pull/251/commits/01322d31c7bf0869cd52aa2ae8e12e35013699f4, but to my surprise it actually passes. `{ isLoading: true, result: TResult }` passes the condition in L69 because it is a logical OR: _not loading OR not empty_ (in this case it is _not empty_), so it is not ignored but shown, but it _will_ (rightfully) not show the loader if it is the latest result at the time the loader delay elapses, because L64 makes sure to filter out non-empty results (_is loading AND not empty_ => emit `loading`, otherwise not). Does this make sense?

I'm really happy you asked about this, because it uncovered a missing test case! Thank you for being so thorough!

Reply:
Oh! I misread the logical OR for a logical AND 😅 This makes sense, and the use case of intermediate results is most definitely valid (and in fact I was curious how this would work _without_ handling intermediate results in the other PR). Thanks for adding the missing tests!